contracts/Depository.sol

```diff
@@ -226,56 +229,71 @@ contract Depository is IErrorsTokenomics {
+        emit CreateProduct(token, productId, supply, priceLP, vesting);
     }
 
     /// @dev Closes bonding products.
     /// @notice This will terminate programs regardless of their vesting time.
     /// @param productIds Set of product Ids.
+    /// @return closedProductIds Set of closed product Ids.
     /// #if_succeeds {:msg "productCounter not touched"} productCounter == old(productCounter);
-    /// #if_succeeds {:msg "success closed"} forall (uint k in productIds) mapBondProducts[productIds[k]].expiry == 0 && mapBondProducts[productIds[k]].supply == 0;
-    function close(uint256[] memory productIds) external {
+    /// #if_succeeds {:msg "success closed"} forall (uint k in productIds) mapBondProducts[productIds[k]].vesting == 0 && mapBondProducts[productIds[k]].supply == 0;
+    function close(uint256[] memory productIds) external returns (uint256[] memory closedProductIds) {
         // Check for the contract ownership
         if (msg.sender != owner) {
             revert OwnerOnly(msg.sender, owner);
         }
 
-        for (uint256 i = 0; i < productIds.length; ++i) {
+        // Calculate the number of closed products
+        uint256 numProducts = productIds.length;
+        bool[] memory positions = new bool[](numProducts);
+        uint256 numClosedProducts;
+        // Traverse to close all possible products
+        for (uint256 i = 0; i < numProducts; ++i) {
             uint256 productId = productIds[i];
-            // Check if the product is still open
-            if (mapBondProducts[productId].expiry == 0) {
-                revert ProductClosed(productId);
-            }
-
+            // Check if the product is still open by getting its supply amount
             uint256 supply = mapBondProducts[productId].supply;
-            // Refund unused OLAS supply from the program if it was not used by the product completely
+            // The supply is greater than zero only if the product is active, otherwise it is instantly closed
             if (supply > 0) {
+                // Refund unused OLAS supply from the product if it was not used by the product completely
                 ITokenomics(tokenomics).refundFromBondProgram(supply);
+                address token = mapBondProducts[productId].token;
+                delete mapBondProducts[productId];
+
+                positions[i] = true;
+                ++numClosedProducts;
+                emit CloseProduct(token, productId, supply);
             }
-            address token = mapBondProducts[productId].token;
-            delete mapBondProducts[productId];
+        }
 
-            emit CloseProduct(token, productId);
+        // Get indexes of closed products
+        closedProductIds = new uint256[](numClosedProducts);
+        uint256 numPos;
+        for (uint256 i = 0; i < numProducts; ++i) {
+            if (positions[i] ) {
+                closedProductIds[numPos] = productIds[i];
+                ++numPos;
+            }
         }
```

Comment:
We need to allocate the resulting array of a correct size, that's why we calculate `numClosedProducts` first. But yes, this can be optimized further such that Ids are recorded in the same loop, and then just written to the array with the correct size. Will raise a PR.